File: go/test/coordinate_conversion_entity_test.go
package sdktest

import (
	"encoding/json"
	"os"
	"path/filepath"
	"runtime"
	"strings"
	"testing"
	"time"

	sdk "github.com/voxgig-sdk/bng2latlong-converter-sdk"
	"github.com/voxgig-sdk/bng2latlong-converter-sdk/core"

	vs "github.com/voxgig/struct"
)

func TestCoordinateConversionEntity(t *testing.T) {
	t.Run("instance", func(t *testing.T) {
		testsdk := sdk.TestSDK(nil, nil)
		ent := testsdk.CoordinateConversion(nil)
		if ent == nil {
			t.Fatal("expected non-nil CoordinateConversionEntity")
		}
	})

	t.Run("basic", func(t *testing.T) {
		setup := coordinate_conversionBasicSetup(nil)
		// Per-op sdk-test-control.json skip — basic test exercises a flow
		// with multiple ops; skipping any op skips the whole flow.
		_mode := "unit"
		if setup.live {
			_mode = "live"
		}
		for _, _op := range []string{"load"} {
			if _shouldSkip, _reason := isControlSkipped("entityOp", "coordinate_conversion." + _op, _mode); _shouldSkip {
				if _reason == "" {
					_reason = "skipped via sdk-test-control.json"
				}
				t.Skip(_reason)
				return
			}
		}
		// The basic flow consumes synthetic IDs from the fixture. In live mode
		// without an *_ENTID env override, those IDs hit the live API and 4xx.
		if setup.syntheticOnly {
			t.Skip("live entity test uses synthetic IDs from fixture — set BNG_LATLONGCONVERTER_TEST_COORDINATE_CONVERSION_ENTID JSON to run live")
			return
		}
		client := setup.client

		// Bootstrap entity data from existing test data (no create step in flow).
		coordinateConversionRef01DataRaw := vs.Items(core.ToMapAny(vs.GetPath("existing.coordinate_conversion", setup.data)))
		var coordinateConversionRef01Data map[string]any
		if len(coordinateConversionRef01DataRaw) > 0 {
			coordinateConversionRef01Data = core.ToMapAny(coordinateConversionRef01DataRaw[0][1])
		}
		// Discard guards against Go's unused-var check when the flow's steps
		// happen not to consume the bootstrap data (e.g. list-only flows).
		_ = coordinateConversionRef01Data

		// LOAD
		coordinateConversionRef01Ent := client.CoordinateConversion(nil)
		coordinateConversionRef01MatchDt0 := map[string]any{}
		coordinateConversionRef01DataDt0Loaded, err := coordinateConversionRef01Ent.Load(coordinateConversionRef01MatchDt0, nil)
		if err != nil {
			t.Fatalf("load failed: %v", err)
		}
		if coordinateConversionRef01DataDt0Loaded == nil {
			t.Fatal("expected load result to be non-nil")
		}

	})
}

func coordinate_conversionBasicSetup(extra map[string]any) *entityTestSetup {
	loadEnvLocal()

	_, filename, _, _ := runtime.Caller(0)
	dir := filepath.Dir(filename)

	entityDataFile := filepath.Join(dir, "..", "..", ".sdk", "test", "entity", "coordinate_conversion", "CoordinateConversionTestData.json")

	entityDataSource, err := os.ReadFile(entityDataFile)
	if err != nil {
		panic("failed to read coordinate_conversion test data: " + err.Error())
	}

	var entityData map[string]any
	if err := json.Unmarshal(entityDataSource, &entityData); err != nil {
		panic("failed to parse coordinate_conversion test data: " + err.Error())
	}

	options := map[string]any{}
	options["entity"] = entityData["existing"]

	client := sdk.TestSDK(options, extra)

	// Generate idmap via transform, matching TS pattern.
	idmap := vs.Transform(
		[]any{"coordinate_conversion01", "coordinate_conversion02", "coordinate_conversion03", "bng2latlong01", "bng2latlong02", "bng2latlong03", "easting01"},
		map[string]any{
			"`$PACK`": []any{"", map[string]any{
				"`$KEY`": "`$COPY`",
				"`$VAL`": []any{"`$FORMAT`", "upper", "`$COPY`"},
			}},
		},
	)

	// Detect ENTID env override before envOverride consumes it. When live
	// mode is on without a real override, the basic test runs against synthetic
	// IDs from the fixture and 4xx's. Surface this so the test can skip.
	entidEnvRaw := os.Getenv("BNG_LATLONGCONVERTER_TEST_COORDINATE_CONVERSION_ENTID")
	idmapOverridden := entidEnvRaw != "" && strings.HasPrefix(strings.TrimSpace(entidEnvRaw), "{")

	env := envOverride(map[string]any{
		"BNG_LATLONGCONVERTER_TEST_COORDINATE_CONVERSION_ENTID": idmap,
		"BNG_LATLONGCONVERTER_TEST_LIVE":      "FALSE",
		"BNG_LATLONGCONVERTER_TEST_EXPLAIN":   "FALSE",
		"BNG_LATLONGCONVERTER_APIKEY":         "NONE",
	})

	idmapResolved := core.ToMapAny(env["BNG_LATLONGCONVERTER_TEST_COORDINATE_CONVERSION_ENTID"])
	if idmapResolved == nil {
		idmapResolved = core.ToMapAny(idmap)
	}

	if env["BNG_LATLONGCONVERTER_TEST_LIVE"] == "TRUE" {
		mergedOpts := vs.Merge([]any{
			map[string]any{
				"apikey": env["BNG_LATLONGCONVERTER_APIKEY"],
			},
			extra,
		})
		client = sdk.NewBng2latlongConverterSDK(core.ToMapAny(mergedOpts))
	}

	live := env["BNG_LATLONGCONVERTER_TEST_LIVE"] == "TRUE"
	return &entityTestSetup{
		client:        client,
		data:          entityData,
		idmap:         idmapResolved,
		env:           env,
		explain:       env["BNG_LATLONGCONVERTER_TEST_EXPLAIN"] == "TRUE",
		live:          live,
		syntheticOnly: live && !idmapOverridden,
		now:           time.Now().UnixMilli(),
	}
}

File: php/test/CoordinateConversionEntityTest.php
<?php
declare(strict_types=1);

// CoordinateConversion entity test

require_once __DIR__ . '/../bng2latlongconverter_sdk.php';
require_once __DIR__ . '/Runner.php';

use PHPUnit\Framework\TestCase;
use Voxgig\Struct\Struct as Vs;

class CoordinateConversionEntityTest extends TestCase
{
    public function test_create_instance(): void
    {
        $testsdk = Bng2latlongConverterSDK::test(null, null);
        $ent = $testsdk->CoordinateConversion(null);
        $this->assertNotNull($ent);
    }

    public function test_basic_flow(): void
    {
        $setup = coordinate_conversion_basic_setup(null);
        // Per-op sdk-test-control.json skip.
        $_live = !empty($setup["live"]);
        foreach (["load"] as $_op) {
            [$_shouldSkip, $_reason] = Runner::is_control_skipped("entityOp", "coordinate_conversion." . $_op, $_live ? "live" : "unit");
            if ($_shouldSkip) {
                $this->markTestSkipped($_reason ?? "skipped via sdk-test-control.json");
                return;
            }
        }
        // The basic flow consumes synthetic IDs from the fixture. In live mode
        // without an *_ENTID env override, those IDs hit the live API and 4xx.
        if (!empty($setup["synthetic_only"])) {
            $this->markTestSkipped("live entity test uses synthetic IDs from fixture — set BNG_LATLONGCONVERTER_TEST_COORDINATE_CONVERSION_ENTID JSON to run live");
            return;
        }
        $client = $setup["client"];

        // Bootstrap entity data from existing test data.
        $coordinate_conversion_ref01_data_raw = Vs::items(Helpers::to_map(
            Vs::getpath($setup["data"], "existing.coordinate_conversion")));
        $coordinate_conversion_ref01_data = null;
        if (count($coordinate_conversion_ref01_data_raw) > 0) {
            $coordinate_conversion_ref01_data = Helpers::to_map($coordinate_conversion_ref01_data_raw[0][1]);
        }

        // LOAD
        $coordinate_conversion_ref01_ent = $client->CoordinateConversion(null);
        $coordinate_conversion_ref01_match_dt0 = [];
        [$coordinate_conversion_ref01_data_dt0_loaded, $err] = $coordinate_conversion_ref01_ent->load($coordinate_conversion_ref01_match_dt0, null);
        $this->assertNull($err);
        $this->assertNotNull($coordinate_conversion_ref01_data_dt0_loaded);

    }
}

function coordinate_conversion_basic_setup($extra)
{
    Runner::load_env_local();

    $entity_data_file = __DIR__ . '/../../.sdk/test/entity/coordinate_conversion/CoordinateConversionTestData.json';
    $entity_data_source = file_get_contents($entity_data_file);
    $entity_data = json_decode($entity_data_source, true);

    $options = [];
    $options["entity"] = $entity_data["existing"];

    $client = Bng2latlongConverterSDK::test($options, $extra);

    // Generate idmap.
    $idmap = [];
    foreach (["coordinate_conversion01", "coordinate_conversion02", "coordinate_conversion03", "bng2latlong01", "bng2latlong02", "bng2latlong03", "easting01"] as $k) {
        $idmap[$k] = strtoupper($k);
    }

    // Detect ENTID env override before envOverride consumes it. When live
    // mode is on without a real override, the basic test runs against synthetic
    // IDs from the fixture and 4xx's. Surface this so the test can skip.
    $entid_env_raw = getenv("BNG_LATLONGCONVERTER_TEST_COORDINATE_CONVERSION_ENTID");
    $idmap_overridden = $entid_env_raw !== false && str_starts_with(trim($entid_env_raw), "{");

    $env = Runner::env_override([
        "BNG_LATLONGCONVERTER_TEST_COORDINATE_CONVERSION_ENTID" => $idmap,
        "BNG_LATLONGCONVERTER_TEST_LIVE" => "FALSE",
        "BNG_LATLONGCONVERTER_TEST_EXPLAIN" => "FALSE",
        "BNG_LATLONGCONVERTER_APIKEY" => "NONE",
    ]);

    $idmap_resolved = Helpers::to_map(
        $env["BNG_LATLONGCONVERTER_TEST_COORDINATE_CONVERSION_ENTID"]);
    if ($idmap_resolved === null) {
        $idmap_resolved = Helpers::to_map($idmap);
    }

    if ($env["BNG_LATLONGCONVERTER_TEST_LIVE"] === "TRUE") {
        $merged_opts = Vs::merge([
            [
                "apikey" => $env["BNG_LATLONGCONVERTER_APIKEY"],
            ],
            $extra ?? [],
        ]);
        $client = new Bng2latlongConverterSDK(Helpers::to_map($merged_opts));
    }

    $live = $env["BNG_LATLONGCONVERTER_TEST_LIVE"] === "TRUE";
    return [
        "client" => $client,
        "data" => $entity_data,
        "idmap" => $idmap_resolved,
        "env" => $env,
        "explain" => $env["BNG_LATLONGCONVERTER_TEST_EXPLAIN"] === "TRUE",
        "live" => $live,
        "synthetic_only" => $live && !$idmap_overridden,
        "now" => (int)(microtime(true) * 1000),
    ];
}
